src/sdam/topology.rs

```diff
@@ -319,7 +228,21 @@ impl Topology {
 #[derive(Debug, Clone)]
 pub(crate) struct TopologyState {
     pub(crate) description: TopologyDescription,
-    pub(crate) servers: HashMap<ServerAddress, Arc<Server>>,
+    servers: HashMap<ServerAddress, Weak<Server>>,
```

Comment:
Storing `Arc<Server>` here actually clones them for every receiver (aka `TopologyWatcher`), which led to a reference cycle wherein the `Server`s were never actually getting dropped because the monitors held watchers, but the monitors never exited because the `Server`s never dropped. Storing `Weak` pointers here fixes this.

One future area to explore would be to ditch `Arc` / `Weak` here altogether and just pass clones of `Server` around, but I didn't want to go too deep down the refactor rabbit hole.

2.3.0 (which has the Topology refactor) is not affected by this issue, but that's because the monitors exit once they detect the topology is dropped, which then causes the `Server`s to drop. They have no need for the `Server` to be alive in order to be dropped. This does highlight a separate issue, which is that monitors on 2.3.0 (and probably earlier) don't exit if a server is removed from the Topology. This is fixed now, however.